p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,404 @@
+package holepunch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	pb "github.com/libp2p/go-libp2p/p2p/protocol/holepunch/pb"
+	"github.com/libp2p/go-libp2p/p2p/protocol/identify"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-msgio/protoio"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+// Protocol is the libp2p protocol for Hole Punching.
+const Protocol protocol.ID = "/libp2p/dcutr"
+
+// StreamTimeout is the timeout for the hole punch protocol stream.
+var StreamTimeout = 1 * time.Minute
+
+// TODO Should we have options for these ?
+const (
+	maxMsgSize  = 4 * 1024 // 4K
+	dialTimeout = 5 * time.Second
+	maxRetries  = 3
+	retryWait   = 2 * time.Second
+)
+
+var (
+	log = logging.Logger("p2p-holepunch")
+	// ErrHolePunchActive is returned from DirectConnect when another hole punching attempt is currently running
+	ErrHolePunchActive = errors.New("another hole punching attempt to this peer is active")
+	// ErrClosed is returned when the hole punching is closed
+	ErrClosed = errors.New("hole punching service closing")
+)
+
+// The Service is used to make direct connections with a peer via hole-punching.
+type Service struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	ids  *identify.IDService
+	host host.Host
+
+	tracer *tracer
+
+	closeMx  sync.RWMutex
+	closed   bool
+	refCount sync.WaitGroup
+
+	// active hole punches for deduplicating
+	activeMx sync.Mutex
+	active   map[peer.ID]struct{}
+}
+
+type Option func(*Service) error
+
+// NewService creates a new service that can be used for hole punching
+func NewService(h host.Host, ids *identify.IDService, opts ...Option) (*Service, error) {
+	if ids == nil {
+		return nil, errors.New("identify service can't be nil")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	hs := &Service{
+		ctx:       ctx,
+		ctxCancel: cancel,
+		host:      h,
+		ids:       ids,
+		active:    make(map[peer.ID]struct{}),
+	}
+
+	for _, opt := range opts {
+		if err := opt(hs); err != nil {
+			cancel()
+			return nil, err
+		}
+	}
+
+	h.SetStreamHandler(Protocol, hs.handleNewStream)
+	h.Network().Notify((*netNotifiee)(hs))
+	return hs, nil
+}
+
+// Close closes the Hole Punch Service.
+func (hs *Service) Close() error {
+	hs.closeMx.Lock()
+	hs.closed = true
```

Comment:
I agree. My point is that the tracer's close function could get called multiple times, and it may not be expecting that.

Of course, it should just "deal" with it, so we're probably fine.